### PR TITLE
Fix noexcept clauses (#37560)

### PR DESCRIPTION
**Pull request by tornaria**
In #36507 I added a lot of `noexcept` clauses guided by a warning which is not quite right (see https://github.com/cython/cython/pull/5999#issuecomment-1986868208).

A new warning in 3.0.9 shows the mistake (incorrectly added `noexcept` clauses). This broke some doctests (#37560) and a workaround was implemented for 10.3 (#37583) since we were too close to release.

This PR now does the proper fix, removing all the incorrect `noexcept`, and also adding a few missing `noexcept`.

Note: if one tries this PR with cython <= 3.0.8 it seems it's wrong in the sense that it will show 40k more warnings after the PR. These are *incorrect* warnings.

If one uses cython 3.0.9 each of these 40k lines give a correct warning before this PR and an incorrect warnings after the PR, and there is no way to avoid 40k warnings.

To confirm this PR is a good one, one needs to use cython 3.0.9 + https://github.com/cython/cython/pull/6087 or wait for cython 3.0.10. In this case, the warnings are correct and one will get 40k warnings before the PR and no warning after the PR.

The last commit reverts the workaround from #37583 so we don't silence these warnings on `cython()` now that we have our own code right.

@vbraun: this is "almost trivial" but touches too many files. Could we merge it in beta0 to avoid conflicts? I did this "kind of" automatically and I've been testing it for two weeks in almost all my builds. The changes made should not affect behaviour at all in the legacy mode we are using cython.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.

### CODE REVIEW

File: src/sage/modular/modform/eis_series_cython.pyx [hunk 1]
```diff
@@ -140,7 +140,7 @@ cpdef Ek_ZZ(int k, int prec=10) noexcept:
     return val
 
 
-cpdef eisenstein_series_poly(int k, int prec = 10)  noexcept:
+cpdef eisenstein_series_poly(int k, int prec = 10) :
```
Comment (mkoeppe) [inline src/sage/modular/modform/eis_series_cython.pyx]:
```suggestion
cpdef eisenstein_series_poly(int k, int prec=10):
```

Comment (tornaria) [reply src/sage/modular/modform/eis_series_cython.pyx]: I didn't make any change other than adding or removing `noexcept`.

We can add style changes all together in a single commit, or do you prefer to fixup the original commit?

Is there an easy way to run a linter/pycodestyle just in the lines that changed?

Comment (mkoeppe) [reply src/sage/modular/modform/eis_series_cython.pyx]: only one so far Screenshot 2024-03-25 at 9 31 59 AM

File: src/sage/numerical/backends/cvxpy_backend.pyx [hunk 2]
```diff
@@ -869,7 +869,7 @@ cdef class CVXPYBackend:
         """
         return self.variables[index].name()
 
-    cpdef variable_upper_bound(self, int index, value = False) noexcept:
+    cpdef variable_upper_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/cvxpy_backend.pyx]:
```suggestion
    cpdef variable_upper_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/cvxpy_backend.pyx [hunk 3]
```diff
@@ -901,7 +901,7 @@ cdef class CVXPYBackend:
         else:
             return self.col_upper_bound[index]
 
-    cpdef variable_lower_bound(self, int index, value = False) noexcept:
+    cpdef variable_lower_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/cvxpy_backend.pyx]:
```suggestion
    cpdef variable_lower_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/generic_backend.pxd [hunk 4]
```diff
@@ -12,49 +12,49 @@ from sage.structure.sage_object cimport SageObject
...
-    cpdef base_ring(self) noexcept
-    cpdef __copy__(self) noexcept
-    cpdef copy(self) noexcept
+    cpdef problem_name(self, name = *)
+    cpdef row_bounds(self, int index)
+    cpdef col_bounds(self, int index)
+    cpdef row_name(self, int index)
+    cpdef col_name(self, int index)
+    cpdef variable_upper_bound(self, int index, value = *)
+    cpdef variable_lower_bound(self, int index, value = *)
```
Comment (mkoeppe) [inline src/sage/numerical/backends/generic_backend.pxd]:
```suggestion
    cpdef variable_upper_bound(self, int index, value=*)
    cpdef variable_lower_bound(self, int index, value=*)
```

File: src/sage/numerical/backends/generic_backend.pxd [hunk 5]
```diff
@@ -12,49 +12,49 @@ from sage.structure.sage_object cimport SageObject
...
+    cpdef copy(self)
     cpdef bint is_variable_basic(self, int index) noexcept
     cpdef bint is_variable_nonbasic_at_lower_bound(self, int index) noexcept
     cpdef bint is_slack_variable_basic(self, int index) noexcept
     cpdef bint is_slack_variable_nonbasic_at_lower_bound(self, int index) noexcept
 
     cdef object obj_constant_term
 
-cpdef GenericBackend get_solver(constraint_generation = ?, solver = ?, base_ring = ?) noexcept
+cpdef GenericBackend get_solver(constraint_generation = ?, solver = ?, base_ring = ?)
```
Comment (mkoeppe) [inline src/sage/numerical/backends/generic_backend.pxd]:
```suggestion
cpdef GenericBackend get_solver(constraint_generation=?, solver=?, base_ring=?)
```

File: src/sage/numerical/backends/generic_backend.pxd [hunk 6]
```diff
@@ -12,49 +12,49 @@ from sage.structure.sage_object cimport SageObject
...
-    cpdef row_name(self, int index) noexcept
-    cpdef col_name(self, int index) noexcept
-    cpdef variable_upper_bound(self, int index, value = *) noexcept
-    cpdef variable_lower_bound(self, int index, value = *) noexcept
-    cpdef solver_parameter(self, name, value=*) noexcept
-    cpdef zero(self) noexcept
-    cpdef base_ring(self) noexcept
-    cpdef __copy__(self) noexcept
-    cpdef copy(self) noexcept
+    cpdef problem_name(self, name = *)
```
Comment (mkoeppe) [inline src/sage/numerical/backends/generic_backend.pxd]:
```suggestion
    cpdef problem_name(self, name=*)
```

File: src/sage/numerical/backends/generic_backend.pyx [hunk 7]
```diff
@@ -335,7 +335,7 @@ cdef class GenericBackend:
         else:
             self.obj_constant_term = d
 
-    cpdef set_objective(self, list coeff, d = 0.0) noexcept:
+    cpdef set_objective(self, list coeff, d = 0.0):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/generic_backend.pyx]:
```suggestion
    cpdef set_objective(self, list coeff, d=0.0):
```

File: src/sage/numerical/backends/generic_backend.pyx [hunk 8]
```diff
@@ -1347,7 +1347,7 @@ cdef class GenericBackend:
         p._test_copy(**options)
         p._test_copy_does_not_share_data(**options)
 
-    cpdef variable_upper_bound(self, int index, value = False) noexcept:
+    cpdef variable_upper_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/generic_backend.pyx]:
```suggestion
    cpdef variable_upper_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/generic_backend.pyx [hunk 9]
```diff
@@ -1374,7 +1374,7 @@ cdef class GenericBackend:
         """
         raise NotImplementedError()
 
-    cpdef variable_lower_bound(self, int index, value = False) noexcept:
+    cpdef variable_lower_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/generic_backend.pyx]:
```suggestion
    cpdef variable_lower_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/generic_backend.pyx [hunk 10]
```diff
@@ -1401,7 +1401,7 @@ cdef class GenericBackend:
         """
         raise NotImplementedError()
 
-    cpdef solver_parameter(self, name, value = None) noexcept:
+    cpdef solver_parameter(self, name, value = None):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/generic_backend.pyx]:
```suggestion
    cpdef solver_parameter(self, name, value=None):
```

File: src/sage/numerical/backends/generic_sdp_backend.pyx [hunk 11]
```diff
@@ -590,7 +590,7 @@ cdef class GenericSDPBackend:
         """
         raise NotImplementedError()
 
-    cpdef solver_parameter(self, name, value = None) noexcept:
+    cpdef solver_parameter(self, name, value = None):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/generic_sdp_backend.pyx]:
```suggestion
    cpdef solver_parameter(self, name, value=None):
```

File: src/sage/numerical/backends/glpk_backend.pxd [hunk 12]
```diff
@@ -26,15 +26,15 @@ cdef class GLPKBackend(GenericBackend):
     cdef glp_smcp * smcp
     cdef int simplex_or_intopt
     cdef search_tree_data_t search_tree_data
-    cpdef __copy__(self) noexcept
+    cpdef __copy__(self)
     cpdef int print_ranges(self, filename = *) except -1
```
Comment (mkoeppe) [inline src/sage/numerical/backends/glpk_backend.pxd]:
```suggestion
    cpdef int print_ranges(self, filename=*) except -1
```

File: src/sage/numerical/backends/glpk_backend.pyx [hunk 13]
```diff
@@ -379,7 +379,7 @@ cdef class GLPKBackend(GenericBackend):
                 raise ValueError("Problem name for GLPK must not be longer than 255 characters.")
             glp_set_prob_name(self.lp, name)
 
-    cpdef set_objective(self, list coeff, d = 0.0) noexcept:
+    cpdef set_objective(self, list coeff, d = 0.0):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/glpk_backend.pyx]:
```suggestion
    cpdef set_objective(self, list coeff, d=0.0):
```

File: src/sage/numerical/backends/glpk_backend.pyx [hunk 14]
```diff
@@ -1600,7 +1600,7 @@ cdef class GLPKBackend(GenericBackend):
 
         return glp_get_obj_dir(self.lp) == GLP_MAX
 
-    cpdef variable_upper_bound(self, int index, value = False) noexcept:
+    cpdef variable_upper_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/glpk_backend.pyx]:
```suggestion
    cpdef variable_upper_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/glpk_backend.pyx [hunk 15]
```diff
@@ -1699,7 +1699,7 @@ cdef class GLPKBackend(GenericBackend):
                     glp_set_col_bnds(self.lp, index + 1, GLP_DB, min, dvalue)
                 sig_off()
 
-    cpdef variable_lower_bound(self, int index, value = False) noexcept:
+    cpdef variable_lower_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/glpk_backend.pyx]:
```suggestion
    cpdef variable_lower_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/glpk_backend.pyx [hunk 16]
```diff
@@ -1874,7 +1874,7 @@ cdef class GLPKBackend(GenericBackend):
         return p
 
 
-    cpdef solver_parameter(self, name, value = None) noexcept:
+    cpdef solver_parameter(self, name, value = None):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/glpk_backend.pyx]:
```suggestion
    cpdef solver_parameter(self, name, value=None):
```

File: src/sage/numerical/backends/glpk_graph_backend.pxd [hunk 17]
```diff
@@ -27,22 +27,22 @@ ctypedef struct c_a_data:
...
-    cpdef list vertices(self) noexcept
-    cpdef add_edge(self, u, v, dict params = ?) noexcept
-    cpdef __add_edges_sage(self, g) noexcept
-    cpdef list add_edges(self, edges) noexcept
-    cpdef delete_edge(self, u, v, dict params = ?) noexcept
-    cpdef tuple get_edge(self, u, v) noexcept
-    cpdef list edges(self) noexcept
-    cpdef delete_vertex(self, vert) noexcept
-    cpdef delete_vertices(self, list verts) noexcept
+    cpdef add_vertex(self, name = ?)
```
Comment (mkoeppe) [inline src/sage/numerical/backends/glpk_graph_backend.pxd]:
```suggestion
    cpdef add_vertex(self, name=?)
```

File: src/sage/numerical/backends/glpk_graph_backend.pxd [hunk 18]
```diff
@@ -27,22 +27,22 @@ ctypedef struct c_a_data:
...
+    cpdef __add_vertices_sage(self, g)
+    cpdef dict get_vertex(self, vertex)
+    cpdef dict get_vertices(self, verts)
+    cpdef set_vertex_demand(self, vertex, param)
+    cpdef set_vertices_demand(self, list pairs)
+    cpdef list vertices(self)
+    cpdef add_edge(self, u, v, dict params = ?)
+    cpdef __add_edges_sage(self, g)
+    cpdef list add_edges(self, edges)
+    cpdef delete_edge(self, u, v, dict params = ?)
```
Comment (mkoeppe) [inline src/sage/numerical/backends/glpk_graph_backend.pxd]:
```suggestion
    cpdef add_edge(self, u, v, dict params=?)
    cpdef __add_edges_sage(self, g)
    cpdef list add_edges(self, edges)
    cpdef delete_edge(self, u, v, dict params=?)
```

File: src/sage/numerical/backends/interactivelp_backend.pyx [hunk 19]
```diff
@@ -396,7 +396,7 @@ cdef class InteractiveLPBackend:
                                            constraint_types, variable_types,
                                            problem_type, ring, objective_constant_term=d)
 
-    cpdef set_objective(self, list coeff, d = 0) noexcept:
+    cpdef set_objective(self, list coeff, d = 0):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/interactivelp_backend.pyx]:
```suggestion
    cpdef set_objective(self, list coeff, d=0):
```

File: src/sage/numerical/backends/interactivelp_backend.pyx [hunk 20]
```diff
@@ -977,7 +977,7 @@ cdef class InteractiveLPBackend:
         """
         return str(self.lp.decision_variables()[index])
 
-    cpdef variable_upper_bound(self, int index, value = False) noexcept:
+    cpdef variable_upper_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/interactivelp_backend.pyx]:
```suggestion
    cpdef variable_upper_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/interactivelp_backend.pyx [hunk 21]
```diff
@@ -1021,7 +1021,7 @@ cdef class InteractiveLPBackend:
                                                constraint_types, variable_types,
                                                problem_type, ring, objective_constant_term=d)
 
-    cpdef variable_lower_bound(self, int index, value = False) noexcept:
+    cpdef variable_lower_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/interactivelp_backend.pyx]:
```suggestion
    cpdef variable_lower_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/ppl_backend.pyx [hunk 22]
```diff
@@ -1063,7 +1063,7 @@ cdef class PPLBackend(GenericBackend):
             return self.col_name_var[index]
         return "x_" + repr(index)
 
-    cpdef variable_upper_bound(self, int index, value = False) noexcept:
+    cpdef variable_upper_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/ppl_backend.pyx]:
```suggestion
    cpdef variable_upper_bound(self, int index, value=False):
```

File: src/sage/numerical/backends/ppl_backend.pyx [hunk 23]
```diff
@@ -1095,7 +1095,7 @@ cdef class PPLBackend(GenericBackend):
         else:
             return self.col_upper_bound[index]
 
-    cpdef variable_lower_bound(self, int index, value = False) noexcept:
+    cpdef variable_lower_bound(self, int index, value = False):
```
Comment (mkoeppe) [inline src/sage/numerical/backends/ppl_backend.pyx]:
```suggestion
    cpdef variable_lower_bound(self, int index, value=False):
```

**Comment by tornaria** [comment]
@mkoeppe do you know if it's possible to get a patch (or patchset) with all the review suggestions?

**Comment by mkoeppe** [comment]
If OK with you, I can just push 1 commit with all of them to your branch

**Comment by mkoeppe** [comment]
By the way, do you know if any of the ones that keep their `noexcept` should be marked `nogil`?

**Comment by tornaria** [comment]
> By the way, do you know if any of the ones that keep their `noexcept` should be marked `nogil`?

Maybe, but that's harder to assess, and more risky. Acquiring the GIL when not needed could affect performance, but not acquiring the GIL when needed could result in crash / deadlock.

**Comment by mkoeppe** [comment]
I agree. Let's hope that a future version of Cython can assist with this

**Comment by mkoeppe** [comment]
These changes all look plausible. Let's merge it.

**Comment by tornaria** [comment]
> These changes all look plausible. Let's merge it.

Thanks.

**Comment by tornaria** [comment]
Summary:
 - The fix for cython has been merged and it will be in 3.0.10 and in 3.1.
 - The fix for numpy has been merged and it will be in 1.26.5 and in 2.0.
 - The fixes for cypari2 and memory_allocator have already been released.
 - With the current PR we should have 0 noexcept warnings left.

**Comment by tornaria** [comment]
The merge I did fixes plural.pxd but plural.pyx has to be fixed as well matching the prototypes.

@vbraun if you disagree that it is better to merge this sooner, could you at least give some feedback?

**Comment by tornaria** [comment]
I rebased on 10.4.beta0 and did the merge properly. The "compare" shows the small difference with respect to the trivial merge done before.

@vbraun could we get this in please?

**Review by mkoeppe — APPROVED** [review]

**Comment by vbraun** [comment]
merge conflict

**Comment by tornaria** [comment]
> merge conflict

@vbraun can you be more specific please?

**Comment by tornaria** [comment]
The merge conflict with #37420 is resolved in the last commit.

**Comment by tornaria** [comment]
Rebased with a single obvious-to-resolve conflict (some code removed in #37420 had one deprecated noexcept clause)

**Pull request merged**